# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/HttpChannelParent.cpp
# Commit: b8f55064bfeb
# Full Hash: b8f55064bfeb151f77c084d04f3def3bcd96679a
# Author: Junior Hsu <juhsu@mozilla.com>
# Date: 2020-06-30 10:00:42
# Regressor Bug: 1633935
# File Overlap Count: 1
# Description:
#   Bug 1633935 - P1 allow OnStartRequest go thru pBg IPC, r=mayhemer
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D73529
# ==============================================================================

diff -r b5ac0f98346f -r b8f55064bfeb netwerk/protocol/http/HttpChannelParent.cpp
--- a/netwerk/protocol/http/HttpChannelParent.cpp	Tue Jun 30 00:31:12 2020 +0000
+++ b/netwerk/protocol/http/HttpChannelParent.cpp	Tue Jun 30 00:32:25 2020 +0000
@@ -1531,14 +1531,29 @@
   }
 
   rv = NS_OK;
-  if (mIPCClosed ||
-      !SendOnStartRequest(
+  bool ipcResult = false;
+  if (!mIPCClosed) {
+    // TODO: For multipart channel, we would deliever everything across
+    // pBackground as well.
+    // TODO: OnStartRequestSent is no longer needed since
+    // OnStartRequest/ODA/OnStopRequest are on the same thread.
+    if (!mIsMultiPart) {
+      ipcResult = mBgParent->OnStartRequest(
           *responseHead, useResponseHead,
           cleanedUpRequest ? cleanedUpRequestHeaders : requestHead->Headers(),
-          args)) {
+          args);
+    } else {
+      ipcResult = SendOnStartRequest(
+          *responseHead, useResponseHead,
+          cleanedUpRequest ? cleanedUpRequestHeaders : requestHead->Headers(),
+          args);
+    }
+  }
+  requestHead->Exit();
+
+  if (mIPCClosed || !ipcResult) {
     rv = NS_ERROR_UNEXPECTED;
   }
-  requestHead->Exit();
 
   // OnStartRequest is sent to content process successfully.
   // Notify PHttpBackgroundChannelChild that all following IPC mesasges